sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala

```diff
@@ -0,0 +1,258 @@
+        updateArray(arrayOfArray.getArray(i).asInstanceOf[GenericArrayData], buffer, inputRow)
+      }
+    }
+    buffer
+  }
+
+  override def merge(buffer: NGramBuffer, input: NGramBuffer): NGramBuffer = {
+    buffer.merge(input)
+    buffer.trim()
+    buffer
+  }
+
+  override def eval(buffer: NGramBuffer): Any = {
+    val topKNGrams = buffer.getTopKNGrams().map((keyValuePair: (Vector[UTF8String], Double)) => {
+      val arrayKey = new GenericArrayData(Vector(new GenericArrayData(keyValuePair._1)))
+      val arrayValue = new GenericArrayData(Vector(keyValuePair._2))
+      new ArrayBasedMapData(arrayKey, arrayValue)
+    }).toVector
+   new GenericArrayData(topKNGrams)
+  }
+
+  private def getNGrams(values: Vector[UTF8String], n: Int): Vector[Vector[UTF8String]] = {
+    if (values.length >= n) {
+      values.sliding(n).toVector
+    }
+    else {
+      Vector()
+    }
+  }
+
+  override def withNewMutableAggBufferOffset(newOffset: Int): NGrams =
+    copy(mutableAggBufferOffset = newOffset)
+
+  override def withNewInputAggBufferOffset(newOffset: Int): NGrams =
+    copy(inputAggBufferOffset = newOffset)
+
+  override def children: Seq[Expression] = Seq(child, nExpression, kExpression, accuracyExpression)
+
+  // Returns null for empty inputs
+  override def nullable: Boolean = true
+
+  override def dataType: DataType = ArrayType(MapType(ArrayType(StringType), DoubleType))
+
+  override def prettyName: String = "ngrams"
+
+  override def serialize(obj: NGramBuffer): Array[Byte] = {
+    NGrams.serializer.serialize(obj)
+  }
+
+  override def deserialize(bytes: Array[Byte]): NGramBuffer = {
+    NGrams.serializer.deserialize(bytes)
+  }
+}
+
+object NGrams {
+
+  val DEFAULT_ACCURACY: Int = 1000
+
+  private def getAccuracy(kExpression: Expression, accuracyExpression: Expression): Expression = {
+    val accuracy = accuracyExpression.eval().asInstanceOf[Int]
+    val k = kExpression.eval().asInstanceOf[Int]
+    Literal(accuracy.max(DEFAULT_ACCURACY / k))
+  }
+
+  val kryoSerializer: KryoSerializer = new KryoSerializer(new SparkConf())
+
+  import collection.JavaConverters._
+
+  class NGramBuffer(val n: Int,
+                    val k: Int,
+                    val precisionFactor: Int,
+                    val frequencyMap: HashMap[Vector[UTF8String], Double]) {
+    def add(ng: Vector[UTF8String]): Unit = {
+      var currentFrequency: Double = frequencyMap.get(ng)
+      if (currentFrequency == null.asInstanceOf[Double]) {
+        currentFrequency = 1.0D
+      } else {
+        currentFrequency += 1
+      }
+      frequencyMap.put(ng, currentFrequency)
+    }
+
+    def merge(other: NGramBuffer): Unit = {
+      other.frequencyMap.asScala.foreach((keyValuePair: (Vector[UTF8String], Double)) => {
+        val key = keyValuePair._1
+        val value = keyValuePair._2
+        val originalValue = frequencyMap.getOrDefault(key, 0.0D)
+        frequencyMap.put(key, originalValue + value)
```

Comment:
Oh. you use java's HashMap.

But `getOrDefault` is only supported in Java8. I think we haven't drop Java7 support for now.